// markdown-exit-shiki/dist/index.js
import { ShikiError as e, codeToHtml as t } from "shiki";
import { mkdir as n, writeFile as r } from "node:fs/promises";
import { dirname as i } from "node:path";
//#region node_modules/@shikijs/colorized-brackets/dist/index.mjs
var a = {
	andromeeda: [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"aurora-x": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"ayu-dark": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"ayu-light": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	"ayu-mirage": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"catppuccin-frappe": [
		"#e78284",
		"#ef9f76",
		"#e5c890",
		"#a6d189",
		"#85c1dc",
		"#ca9ee6",
		"#ea999c"
	],
	"catppuccin-latte": [
		"#d20f39",
		"#fe640b",
		"#df8e1d",
		"#40a02b",
		"#209fb5",
		"#8839ef",
		"#e64553"
	],
	"catppuccin-macchiato": [
		"#ed8796",
		"#f5a97f",
		"#eed49f",
		"#a6da95",
		"#7dc4e4",
		"#c6a0f6",
		"#ee99a0"
	],
	"catppuccin-mocha": [
		"#f38ba8",
		"#fab387",
		"#f9e2af",
		"#a6e3a1",
		"#74c7ec",
		"#cba6f7",
		"#eba0ac"
	],
	"dark-plus": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	dracula: [
		"#F8F8F2",
		"#FF79C6",
		"#8BE9FD",
		"#50FA7B",
		"#BD93F9",
		"#FFB86C",
		"#FF5555"
	],
	"dracula-soft": [
		"#f6f6f4",
		"#f286c4",
		"#97e1f1",
		"#62e884",
		"#bf9eee",
		"#FFB86C",
		"#ee6666"
	],
	"everforest-dark": [
		"#e67e80",
		"#dbbc7f",
		"#a7c080",
		"#7fbbb3",
		"#e69875",
		"#d699b6",
		"#859289"
	],
	"everforest-light": [
		"#f85552",
		"#dfa000",
		"#8da101",
		"#3a94c5",
		"#f57d26",
		"#df69ba",
		"#939f91"
	],
	"github-dark": [
		"#79b8ff",
		"#ffab70",
		"#b392f0",
		"#79b8ff",
		"#ffab70",
		"#b392f0",
		"rgba(255, 18, 18, 0.8)"
	],
	"github-dark-default": [
		"#79c0ff",
		"#56d364",
		"#e3b341",
		"#ffa198",
		"#ff9bce",
		"#d2a8ff",
		"#7d8590"
	],
	"github-dark-dimmed": [
		"#6cb6ff",
		"#6bc46d",
		"#daaa3f",
		"#ff938a",
		"#fc8dc7",
		"#dcbdfb",
		"#768390"
	],
	"github-dark-high-contrast": [
		"#91cbff",
		"#4ae168",
		"#f7c843",
		"#ffb1af",
		"#ffadd4",
		"#dbb7ff",
		"#f0f3f6"
	],
	"github-light": [
		"#005cc5",
		"#e36209",
		"#5a32a3",
		"#005cc5",
		"#e36209",
		"#5a32a3",
		"rgba(255, 18, 18, 0.8)"
	],
	"github-light-default": [
		"#0969da",
		"#1a7f37",
		"#9a6700",
		"#cf222e",
		"#bf3989",
		"#8250df",
		"#656d76"
	],
	"github-light-high-contrast": [
		"#0349b4",
		"#055d20",
		"#744500",
		"#a0111f",
		"#971368",
		"#622cbc",
		"#0e1116"
	],
	"gruvbox-dark-hard": [
		"#b16286",
		"#458588",
		"#689d6a",
		"#98971a",
		"#d79921",
		"#d65d0e",
		"#cc241d"
	],
	"gruvbox-dark-medium": [
		"#b16286",
		"#458588",
		"#689d6a",
		"#98971a",
		"#d79921",
		"#d65d0e",
		"#cc241d"
	],
	"gruvbox-dark-soft": [
		"#b16286",
		"#458588",
		"#689d6a",
		"#98971a",
		"#d79921",
		"#d65d0e",
		"#cc241d"
	],
	"gruvbox-light-hard": [
		"#b16286",
		"#458588",
		"#689d6a",
		"#98971a",
		"#d79921",
		"#d65d0e",
		"#cc241d"
	],
	"gruvbox-light-medium": [
		"#b16286",
		"#458588",
		"#689d6a",
		"#98971a",
		"#d79921",
		"#d65d0e",
		"#cc241d"
	],
	"gruvbox-light-soft": [
		"#b16286",
		"#458588",
		"#689d6a",
		"#98971a",
		"#d79921",
		"#d65d0e",
		"#cc241d"
	],
	horizon: [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"horizon-bright": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	houston: [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"kanagawa-dragon": [
		"#8992A7",
		"#B6927B",
		"#8BA4B0",
		"#A292A3",
		"#C4B28A",
		"#8EA4A2",
		"#C4746E"
	],
	"kanagawa-lotus": [
		"#624C83",
		"#CC6D00",
		"#4D699B",
		"#B35B79",
		"#77713F",
		"#597B75",
		"#D9A594"
	],
	"kanagawa-wave": [
		"#957FB8",
		"#FFA066",
		"#7E9CD8",
		"#D27E99",
		"#E6C384",
		"#7AA89F",
		"#FF5D62"
	],
	laserwave: [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"light-plus": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	"material-theme": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"material-theme-darker": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"material-theme-lighter": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	"material-theme-ocean": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"material-theme-palenight": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"min-dark": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"min-light": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	monokai: [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"night-owl": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"night-owl-light": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	nord: [
		"#8fbcbb",
		"#88c0d0",
		"#81a1c1",
		"#5e81ac",
		"#8fbcbb",
		"#88c0d0",
		"#bf616a"
	],
	"one-dark-pro": [
		"#d19a66",
		"#c678dd",
		"#56b6c2",
		"rgba(255, 18, 18, 0.8)"
	],
	"one-light": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	plastic: [
		"#A9B2C3",
		"#61AFEF",
		"#E5C07B",
		"#E06C75",
		"#98C379",
		"#B57EDC",
		"#D74E42"
	],
	poimandres: [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	red: [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"rose-pine": [
		"#eb6f9280",
		"#31748f80",
		"#f6c17780",
		"#9ccfd880",
		"#ebbcba80",
		"#c4a7e780",
		"rgba(255, 18, 18, 0.8)"
	],
	"rose-pine-dawn": [
		"#b4637a80",
		"#28698380",
		"#ea9d3480",
		"#56949f80",
		"#d7827e80",
		"#907aa980",
		"rgba(255, 18, 18, 0.8)"
	],
	"rose-pine-moon": [
		"#eb6f9280",
		"#3e8fb080",
		"#f6c17780",
		"#9ccfd880",
		"#ea9a9780",
		"#c4a7e780",
		"rgba(255, 18, 18, 0.8)"
	],
	"slack-dark": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"slack-ochin": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	"snazzy-light": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	"solarized-dark": [
		"#cdcdcdff",
		"#b58900ff",
		"#d33682ff",
		"rgba(255, 18, 18, 0.8)"
	],
	"solarized-light": [
		"#0431FA",
		"#319331",
		"#7B3814",
		"rgba(255, 18, 18, 0.8)"
	],
	"synthwave-84": [
		"#FFD700",
		"#DA70D6",
		"#179FFF",
		"rgba(255, 18, 18, 0.8)"
	],
	"tokyo-night": [
		"#698cd6",
		"#68b3de",
		"#9a7ecc",
		"#25aac2",
		"#80a856",
		"#c49a5a",
		"#db4b4b"
	],
	vesper: [
		"#A0A0A0",
		"#A0A0A0",
		"#A0A0A0",
		"#A0A0A0",
		"#A0A0A0",
		"#A0A0A0",
		"#FF8080"
	],
	"vitesse-black": [
		"#5eaab5",
		"#4d9375",
		"#d4976c",
		"#d9739f",
		"#e6cc77",
		"#6394bf",
		"rgba(255, 18, 18, 0.8)"
	],
	"vitesse-dark": [
		"#5eaab5",
		"#4d9375",
		"#d4976c",
		"#d9739f",
		"#e6cc77",
		"#6394bf",
		"rgba(255, 18, 18, 0.8)"
	],
	"vitesse-light": [
		"#2993a3",
		"#1e754f",
		"#a65e2b",
		"#a13865",
		"#bda437",
		"#296aa3",
		"rgba(255, 18, 18, 0.8)"
	]
};
function o(e) {
	return e.explanation?.[0].scopes.findLast((e) => e.scopeName.match(/^source.\w+$/))?.scopeName.split(".")[1];
}
function s(e, t) {
	return {
		themes: e.langs[t]?.themes ?? e.themes,
		bracketPairs: e.langs[t]?.bracketPairs ?? e.bracketPairs
	};
}
function c(e) {
	return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function l(e, t, n) {
	if (!e.explanation) return !0;
	let r = e.explanation?.[0].scopes.findLastIndex((e) => e.scopeName.startsWith("comment.")) ?? -1, i = e.explanation?.[0].scopes.findLastIndex((e) => e.scopeName.startsWith("string.")) ?? -1, a = e.explanation?.[0].scopes.findLastIndex((e) => e.scopeName.startsWith("meta.embedded.") || e.scopeName.startsWith("scope.embedded.") || e.scopeName === "entity.name.type.instance.jsdoc" || e.scopeName === "variable.other.jsdoc" || e.scopeName === "meta.object.liquid") ?? -1;
	return !!(r > a || i > a || t && t.length && !e.explanation?.some((e) => e.scopes.some((e) => t.some((t) => e.scopeName === t || e.scopeName.startsWith(`${t}.`)))) || n && n.length && e.explanation?.some((e) => e.scopes.some((e) => n.some((t) => e.scopeName === t || e.scopeName.startsWith(`${t}.`)))));
}
function u(e, t, n, r) {
	let i = [];
	for (let a of e) {
		let e = s(t, o(a) ?? r), c = new Set(e.bracketPairs.map((e) => e.opener)), u = new Set(e.bracketPairs.map((e) => e.closer)), f = Object.fromEntries(e.bracketPairs.map((e) => [e.closer, e.opener])), p = e.bracketPairs.find((e) => e.opener === a.content.trim() || e.closer === a.content.trim());
		if (!(!p || l(a, p.scopesAllowList, p.scopesDenyList))) {
			if (c.has(a.content.trim())) i.push(a);
			else if (u.has(a.content.trim())) {
				let t = i.slice().reverse().find((e) => e.content.trim() === f[a.content.trim()]);
				if (t) {
					for (; i.at(-1) !== t;) {
						let t = i.pop();
						t && d(t, e.themes, n, -1);
					}
					i.pop(), d(a, e.themes, n, i.length), d(t, e.themes, n, i.length);
				} else d(a, e.themes, n, -1);
			}
		}
	}
	for (let e of i) d(e, s(t, r).themes, n, -1);
}
function d(t, n, r, i) {
	if (f(r)) t.color = m(n, typeof r.theme == "string" ? r.theme : r.theme.name, i);
	else {
		let { defaultColor: a = "light", cssVariablePrefix: o = "--shiki-" } = r, s = typeof t.htmlStyle == "string" ? {} : structuredClone(t.htmlStyle ?? {});
		for (let [e, t] of Object.entries(r.themes)) {
			let r = typeof t == "string" ? t : t?.name, c = e === a ? "color" : `${o}${e}`;
			s[c] = m(n, r, i);
		}
		if (a === "light-dark()") {
			let t = s[`${o}light`], n = s[`${o}dark`];
			if (!t || !n) throw new e("When using `defaultColor: \"light-dark()\"`, you must provide both `light` and `dark` themes");
			s.color = `light-dark(${t},${n})`;
		}
		t.htmlStyle = s;
	}
}
function f(e) {
	return "theme" in e;
}
var p = [
	"#FFD700",
	"#DA70D6",
	"#179FFF",
	"rgba(255, 18, 18, 0.8)"
];
function m(e, t, n) {
	let r = t == null ? p : h(t, e) ?? h(t, a) ?? p;
	return n === -1 ? r[r.length - 1] : r[n % (r.length - 1)];
}
function h(e, t) {
	if (t[e]) return t[e];
	let n = Object.keys(t).sort().reverse().find((t) => e.startsWith(t));
	return n ? t[n] : null;
}
function ee(e, t, n) {
	let r = s(t, o(e) ?? n);
	if (r.bracketPairs.length === 0 || l(e)) return [e];
	let i = new RegExp(r.bracketPairs.flatMap((e) => [e.opener, e.closer]).sort((e, t) => t.length - e.length).map(c).join("|")), a = [e];
	for (;;) {
		let e = a.pop();
		if (!e) break;
		let t = e?.content.match(i);
		if (!t) {
			a.push(e);
			break;
		}
		let n = t.index ?? 0;
		if (n > 0 && a.push({
			...e,
			content: e.content.substring(0, n)
		}), a.push({
			...e,
			content: t[0],
			offset: e.offset + n
		}), n + t[0].length < e.content.length) a.push({
			...e,
			content: e.content.substring(n + t[0].length),
			offset: e.offset + n + t[0].length
		});
		else break;
	}
	let u = e.explanation ?? [], d = 0, f = (u ?? []).map((t, n) => {
		let r = d, i = t.content.length;
		return u.length === 1 ? i = e.content.length : n === 0 ? i = (e.content.match(/^\s*/)?.[0].length ?? 0) + t.content.trimStart().length : n === u.length - 1 && (i = t.content.trimEnd().length + (e.content.match(/\s*$/)?.[0].length ?? 0)), d += i, {
			...t,
			start: r,
			end: r + i - 1
		};
	});
	for (let t of a) {
		let n = t.offset - e.offset, r = n + t.content.length - 1;
		t.explanation = f.filter((e) => n >= e.start && n <= e.end || r >= e.start && r <= e.end || e.start >= n && e.start <= r || e.end >= n && e.end <= r).map((e, t) => u[t]);
	}
	return a;
}
var g = [
	{
		opener: "[",
		closer: "]"
	},
	{
		opener: "{",
		closer: "}"
	},
	{
		opener: "(",
		closer: ")"
	},
	{
		opener: "{{",
		closer: "}}"
	},
	{
		opener: "{%",
		closer: "%}"
	}
];
function _(e = {}) {
	let t = {
		themes: e.themes ?? {},
		bracketPairs: e.bracketPairs ?? [
			{
				opener: "[",
				closer: "]"
			},
			{
				opener: "{",
				closer: "}"
			},
			{
				opener: "(",
				closer: ")"
			},
			{
				opener: "<",
				closer: ">",
				scopesAllowList: [
					"punctuation.definition.typeparameters.begin.ts",
					"punctuation.definition.typeparameters.end.ts",
					"entity.name.type.instance.jsdoc"
				]
			}
		],
		langs: {
			html: { bracketPairs: [] },
			jinja: { bracketPairs: g },
			liquid: { bracketPairs: g },
			...e.langs
		},
		explicitTrigger: e.explicitTrigger ?? !1
	};
	return {
		name: "colorizedBrackets",
		preprocess(e, n) {
			y(t, this.options.meta?.__raw) && (n.includeExplanation ||= "scopeName");
		},
		tokens: function(e) {
			if (!y(t, this.options.meta?.__raw)) return;
			let n = this.options.lang;
			for (let r = 0; r < e.length; r++) e[r] = e[r].flatMap((e) => ee(e, t, n));
			u(e.flat(), t, this.options, n);
		}
	};
}
var v = /(^|\s)colorize-brackets($|\s)/;
function y(e, t) {
	return !e.explicitTrigger || t?.match(v) != null;
}
//#endregion
//#region node_modules/@shikijs/transformers/dist/index.mjs
var b = [
	[/^(<!--)(.+)(-->)$/, !1],
	[/^(\/\*)(.+)(\*\/)$/, !1],
	[/^(\/\/|["'#]|;{1,2}|%{1,2}|--)(.*)$/, !0],
	[/^(\*)(.+)$/, !0]
];
function x(e, t, n) {
	let r = [];
	for (let i of e) {
		if (n === "v3") {
			let e = i.children.flatMap((e, t) => {
				if (e.type !== "element") return e;
				let n = e.children[0];
				if (n.type !== "text") return e;
				let r = t === i.children.length - 1;
				if (!C(n.value, r)) return e;
				let a = n.value.split(/(\s+\/\/)/);
				if (a.length <= 1) return e;
				let o = [a[0]];
				for (let e = 1; e < a.length; e += 2) o.push(a[e] + (a[e + 1] || ""));
				return o = o.filter(Boolean), o.length <= 1 ? e : o.map((t) => ({
					...e,
					children: [{
						type: "text",
						value: t
					}]
				}));
			});
			e.length !== i.children.length && (i.children = e);
		}
		let e = i.children, a = e.length - 1;
		n === "v1" ? a = 0 : t && (a = e.length - 2);
		for (let n = Math.max(a, 0); n < e.length; n++) {
			let a = e[n];
			if (a.type !== "element") continue;
			let o = a.children.at(0);
			if (o?.type !== "text") continue;
			let s = n === e.length - 1, c = C(o.value, s);
			if (!c && n > 0 && o.value.trim().startsWith("[!code")) {
				let t = e[n - 1];
				if (t?.type === "element") {
					let n = t.children.at(0);
					if (n?.type === "text" && n.value.includes("//")) {
						let l = C(n.value + o.value, s);
						if (l) {
							c = l, r.push({
								info: l,
								line: i,
								token: t,
								isLineCommentOnly: e.length === 2 && t.children.length === 1 && a.children.length === 1,
								isJsxStyle: !1,
								additionalTokens: [a]
							});
							continue;
						}
					}
				}
			}
			if (c) if (t && !s && n !== 0) {
				let t = S(e[n - 1], "{") && S(e[n + 1], "}");
				r.push({
					info: c,
					line: i,
					token: a,
					isLineCommentOnly: e.length === 3 && a.children.length === 1,
					isJsxStyle: t,
					additionalTokens: void 0
				});
			} else r.push({
				info: c,
				line: i,
				token: a,
				isLineCommentOnly: e.length === 1 && a.children.length === 1,
				isJsxStyle: !1,
				additionalTokens: void 0
			});
		}
	}
	return r;
}
function S(e, t) {
	if (e.type !== "element") return !1;
	let n = e.children[0];
	return n.type === "text" ? n.value.trim() === t : !1;
}
function C(e, t) {
	let n = e.trimStart(), r = e.length - n.length;
	n = n.trimEnd();
	let i = e.length - n.length - r;
	for (let [e, a] of b) {
		if (a && !t) continue;
		let o = e.exec(n);
		if (o) return [
			" ".repeat(r) + o[1],
			o[2],
			o[3] ? o[3] + " ".repeat(i) : void 0
		];
	}
}
function w(e) {
	let t = e.match(/(?:\/\/|["'#]|;{1,2}|%{1,2}|--)(\s*)$/);
	return t && t[1].trim().length === 0 ? e.slice(0, t.index) : e;
}
function T(e, t, n, r) {
	return r ??= "v3", {
		name: e,
		code(e) {
			let i = e.children.filter((e) => e.type === "element"), a = [];
			e.data ??= {};
			let o = e.data;
			o._shiki_notation ??= x(i, ["jsx", "tsx"].includes(this.options.lang), r);
			let s = o._shiki_notation;
			for (let e of s) {
				if (e.info[1].length === 0) continue;
				let o = i.indexOf(e.line);
				e.isLineCommentOnly && r !== "v1" && o++;
				let s = !1;
				if (e.info[1] = e.info[1].replace(t, (...t) => n.call(this, t, e.line, e.token, i, o) ? (s = !0, "") : t[0]), !s) continue;
				r === "v1" && (e.info[1] = w(e.info[1]));
				let c = e.info[1].trim().length === 0;
				if (c && (e.info[1] = ""), c && e.isLineCommentOnly) a.push(e.line);
				else if (c && e.isJsxStyle) e.line.children.splice(e.line.children.indexOf(e.token) - 1, 3);
				else if (c) {
					if (e.additionalTokens) for (let t = e.additionalTokens.length - 1; t >= 0; t--) {
						let n = e.additionalTokens[t], r = e.line.children.indexOf(n);
						r !== -1 && e.line.children.splice(r, 1);
					}
					e.line.children.splice(e.line.children.indexOf(e.token), 1);
				} else {
					let t = e.token.children[0];
					if (t.type === "text" && (t.value = e.info.join(""), e.additionalTokens)) for (let t of e.additionalTokens) {
						let e = t.children[0];
						e?.type === "text" && (e.value = "");
					}
				}
			}
			for (let t of a) {
				let n = e.children.indexOf(t), r = e.children[n + 1], i = 1;
				r?.type === "text" && r?.value === "\n" && (i = 2), e.children.splice(n, i);
			}
		}
	};
}
function te(e = []) {
	return {
		name: "@shikijs/transformers:compact-line-options",
		line(t, n) {
			let r = e.find((e) => e.line === n);
			return r?.classes && this.addClassToHast(t, r.classes), t;
		}
	};
}
function E(e) {
	if (!e) return null;
	let t = e.match(/\{([\d,-]+)\}/);
	return t ? t[1].split(",").flatMap((e) => {
		let t = e.split("-").map((e) => Number.parseInt(e, 10));
		return t.length === 1 ? [t[0]] : Array.from({ length: t[1] - t[0] + 1 }, (e, n) => t[0] + n);
	}) : null;
}
var D = Symbol("highlighted-lines");
function O(e = {}) {
	let { className: t = "highlighted", zeroIndexed: n = !1 } = e;
	return {
		name: "@shikijs/transformers:meta-highlight",
		line(e, r) {
			if (!this.options.meta?.__raw) return;
			let i = this.meta;
			i[D] ??= E(this.options.meta.__raw);
			let a = i[D] ?? [], o = n ? r - 1 : r;
			return a.includes(o) && this.addClassToHast(e, t), e;
		}
	};
}
function k(e) {
	return e ? Array.from(e.matchAll(/\/((?:\\.|[^/])+)\//g)).map((e) => e[1].replace(/\\(.)/g, "$1")) : [];
}
function A(e = {}) {
	let { className: t = "highlighted-word" } = e;
	return {
		name: "@shikijs/transformers:meta-word-highlight",
		preprocess(e, n) {
			if (!this.options.meta?.__raw) return;
			let r = k(this.options.meta.__raw);
			n.decorations ||= [];
			for (let i of r) {
				let r = j(e, i);
				for (let e of r) n.decorations.push({
					start: e,
					end: e + i.length,
					properties: { class: t }
				});
			}
		}
	};
}
function j(e, t) {
	let n = [], r = 0;
	for (;;) {
		let i = e.indexOf(t, r);
		if (i === -1 || i >= e.length || i < r) break;
		n.push(i), r = i + t.length;
	}
	return n;
}
function M(e) {
	return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function N(e = {}, t = "@shikijs/transformers:notation-map") {
	let { classMap: n = {}, classActivePre: r = void 0, classActiveCode: i = void 0 } = e;
	return T(t, RegExp(`#?\\s*\\[!code (${Object.keys(n).map(M).join("|")})(:\\d+)?\\]`, "gi"), function([e, t, a = ":1"], o, s, c, l) {
		let u = Number.parseInt(a.slice(1), 10);
		for (let e = l; e < Math.min(l + u, c.length); e++) this.addClassToHast(c[e], n[t]);
		return r && this.addClassToHast(this.pre, r), i && this.addClassToHast(this.code, i), !0;
	}, e.matchAlgorithm);
}
function P(e = {}) {
	let { classLineAdd: t = "diff add", classLineRemove: n = "diff remove", classActivePre: r = "has-diff", classActiveCode: i } = e;
	return N({
		classMap: {
			"++": t,
			"--": n
		},
		classActivePre: r,
		classActiveCode: i,
		matchAlgorithm: e.matchAlgorithm
	}, "@shikijs/transformers:notation-diff");
}
function F(e = {}) {
	let { classMap: t = {
		error: ["highlighted", "error"],
		warning: ["highlighted", "warning"],
		info: ["highlighted", "info"]
	}, classActivePre: n = "has-highlighted", classActiveCode: r } = e;
	return N({
		classMap: t,
		classActivePre: n,
		classActiveCode: r,
		matchAlgorithm: e.matchAlgorithm
	}, "@shikijs/transformers:notation-error-level");
}
function I(e = {}) {
	let { classActiveLine: t = "focused", classActivePre: n = "has-focused", classActiveCode: r } = e;
	return N({
		classMap: { focus: t },
		classActivePre: n,
		classActiveCode: r,
		matchAlgorithm: e.matchAlgorithm
	}, "@shikijs/transformers:notation-focus");
}
function L(e = {}) {
	let { classActiveLine: t = "highlighted", classActivePre: n = "has-highlighted", classActiveCode: r } = e;
	return N({
		classMap: {
			highlight: t,
			hl: t
		},
		classActivePre: n,
		classActiveCode: r,
		matchAlgorithm: e.matchAlgorithm
	}, "@shikijs/transformers:notation-highlight");
}
function R(e, t, n, r) {
	let i = z(e), a = i.indexOf(n);
	for (; a !== -1;) B.call(this, e.children, t, a, n.length, r), a = i.indexOf(n, a + 1);
}
function z(e) {
	return e.type === "text" ? e.value : e.type === "element" && e.tagName === "span" ? e.children.map(z).join("") : "";
}
function B(e, t, n, r, i) {
	let a = 0;
	for (let o = 0; o < e.length; o++) {
		let s = e[o];
		if (s.type !== "element" || s.tagName !== "span" || s === t) continue;
		let c = s.children[0];
		if (c.type === "text") {
			if (V([a, a + c.value.length - 1], [n, n + r])) {
				let t = Math.max(0, n - a), l = r - Math.max(0, a - n);
				if (l === 0) continue;
				let u = ne(s, c, t, l);
				this.addClassToHast(u[1], i);
				let d = u.filter(Boolean);
				e.splice(o, 1, ...d), o += d.length - 1;
			}
			a += c.value.length;
		}
	}
}
function V(e, t) {
	return e[0] <= t[1] && e[1] >= t[0];
}
function ne(e, t, n, r) {
	let i = t.value, a = (t) => H(e, { children: [{
		type: "text",
		value: t
	}] });
	return [
		n > 0 ? a(i.slice(0, n)) : void 0,
		a(i.slice(n, n + r)),
		n + r < i.length ? a(i.slice(n + r)) : void 0
	];
}
function H(e, t) {
	return {
		...e,
		properties: { ...e.properties },
		...t
	};
}
function U(e = {}) {
	let { classActiveWord: t = "highlighted-word", classActivePre: n = void 0 } = e;
	return T("@shikijs/transformers:notation-highlight-word", /\s*\[!code word:((?:\\.|[^:\]])+)(:\d+)?\]/, function([e, r, i], a, o, s, c) {
		let l = i ? Number.parseInt(i.slice(1), 10) : s.length;
		r = r.replace(/\\(.)/g, "$1");
		for (let e = c; e < Math.min(c + l, s.length); e++) R.call(this, s[e], o, r, t);
		return n && this.addClassToHast(this.pre, n), !0;
	}, e.matchAlgorithm);
}
function W() {
	return {
		name: "@shikijs/transformers:remove-line-break",
		code(e) {
			e.children = e.children.filter((e) => !(e.type === "text" && e.value === "\n"));
		}
	};
}
function G() {
	return {
		name: "@shikijs/transformers:remove-notation-escape",
		code(e) {
			function t(e) {
				if (e.type === "text") e.value = e.value.replace("[\\!code", "[!code");
				else if ("children" in e) for (let n of e.children) t(n);
			}
			return t(e), e;
		}
	};
}
function K(e) {
	return e === "	";
}
function q(e) {
	return e === " " || e === "	";
}
function J(e) {
	let t = [], n = "";
	function r() {
		n.length && t.push(n), n = "";
	}
	return e.forEach((i, a) => {
		K(i) || q(i) && (q(e[a - 1]) || q(e[a + 1])) ? (r(), t.push(i)) : n += i;
	}), r(), t;
}
function Y(e, t, n = !0) {
	if (t === "all") return e;
	let r = 0, i = 0;
	if (t === "boundary" || t === "leading") for (let t = 0; t < e.length && q(e[t]); t++) r++;
	if (t === "boundary" || t === "trailing") for (let t = e.length - 1; t >= 0 && q(e[t]); t--) i++;
	let a = e.slice(r, e.length - i);
	return [
		...e.slice(0, r),
		...n ? J(a) : [a.join("")],
		...e.slice(e.length - i)
	];
}
function X(e = {}) {
	let t = {
		" ": e.classSpace ?? "space",
		"	": e.classTab ?? "tab"
	}, n = e.position ?? "all", r = Object.keys(t);
	return {
		name: "@shikijs/transformers:render-whitespace",
		root(e) {
			let i = e.children[0];
			(i.tagName === "pre" ? i.children[0] : { children: [e] }).children.forEach((e) => {
				if (e.type !== "element" && e.type !== "root") return;
				let i = e.children.filter((e) => e.type === "element"), a = i.length - 1;
				e.children = e.children.flatMap((e) => {
					if (e.type !== "element") return e;
					let o = i.indexOf(e);
					if (n === "boundary" && o !== 0 && o !== a || n === "trailing" && o !== a || n === "leading" && o !== 0) return e;
					let s = e.children[0];
					if (s.type !== "text" || !s.value) return e;
					let c = Y(s.value.split(/([ \t])/).filter((e) => e.length), n === "boundary" && o === a && a !== 0 ? "trailing" : n, n !== "trailing" && n !== "leading");
					return c.length <= 1 ? e : c.map((n) => {
						let i = {
							...e,
							properties: { ...e.properties }
						};
						return i.children = [{
							type: "text",
							value: n
						}], r.includes(n) && (this.addClassToHast(i, t[n]), delete i.properties.style), i;
					});
				});
			});
		}
	};
}
function Z(e = {}) {
	let { classPrefix: t = "__shiki_", classSuffix: n = "", classReplacer: r = (e) => e } = e, i = /* @__PURE__ */ new Map();
	function a(e) {
		return Object.entries(e).map(([e, t]) => `${e}:${t}`).join(";");
	}
	function o(e) {
		let o = t + re(typeof e == "string" ? e : a(e)) + n;
		return o = r(o), i.has(o) || i.set(o, typeof e == "string" ? e : { ...e }), o;
	}
	return {
		name: "@shikijs/transformers:style-to-class",
		pre(e) {
			if (!e.properties.style) return;
			let t = o(e.properties.style);
			delete e.properties.style, this.addClassToHast(e, t);
		},
		tokens(e) {
			for (let t of e) for (let e of t) {
				if (!e.htmlStyle) continue;
				let t = o(e.htmlStyle);
				e.htmlStyle = {}, e.htmlAttrs ||= {}, e.htmlAttrs.class ? e.htmlAttrs.class += ` ${t}` : e.htmlAttrs.class = t;
			}
		},
		getClassRegistry() {
			return i;
		},
		getCSS() {
			let e = "";
			for (let [t, n] of i.entries()) e += `.${t}{${typeof n == "string" ? n : a(n)}}`;
			return e;
		},
		clearRegistry() {
			i.clear();
		}
	};
}
function re(e, t = 0) {
	let n = 3735928559 ^ t, r = 1103547991 ^ t;
	for (let t = 0, i; t < e.length; t++) i = e.charCodeAt(t), n = Math.imul(n ^ i, 2654435761), r = Math.imul(r ^ i, 1597334677);
	return n = Math.imul(n ^ n >>> 16, 2246822507), n ^= Math.imul(r ^ r >>> 13, 3266489909), r = Math.imul(r ^ r >>> 16, 2246822507), r ^= Math.imul(n ^ n >>> 13, 3266489909), (4294967296 * (2097151 & r) + (n >>> 0)).toString(36).slice(0, 6);
}
//#endregion
//#region src/parse_config.ts
var Q = [
	te(),
	O(),
	A(),
	P(),
	F(),
	I(),
	L(),
	U(),
	W(),
	G(),
	X(),
	_()
];
function ie(e) {
	let t = e || {}, n = [];
	for (let e of t.transformers || ["all"]) {
		e === "all" && (n = Q);
		let t = Q.find((t) => t.name === e);
		t && n.push(t);
	}
	let r = null;
	return t.style_to_class?.enable && (r = Z({ classPrefix: t.style_to_class.class_prefix || "_sk_" }), n.push(r)), {
		themes: t.themes || {
			light: "catppuccin-latte",
			dark: "catppuccin-mocha"
		},
		excludes: t.exclude_languages || ["mermaid"],
		aliases: t.language_aliases || {},
		collapseConfig: (() => {
			let e = t.code_collapse ?? 30;
			return {
				enable: e > 0,
				maxLines: e
			};
		})(),
		styleToClass: {
			enable: t.style_to_class?.enable || !1,
			cssGetter: r?.getCSS,
			css_output_path: t.style_to_class?.css_output_path
		},
		transformers: n,
		toolbarItems: {
			lang: t.toolbar_items?.lang ?? !0,
			title: t.toolbar_items?.title ?? !0,
			wrapToggle: t.toolbar_items?.wrapToggle ?? !0,
			copyButton: t.toolbar_items?.copyButton ?? !0
		}
	};
}
//#endregion
//#region src/shiki_utils.ts
function $(e) {
	return e.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll("\"", "&quot;").replaceAll("'", "&#39;");
}
function ae(e, t, n) {
	let r = ["<div class=\"left\">", "<div class=\"traffic-lights\"> <span class=\"traffic-light red\"></span> <span class=\"traffic-light yellow\"></span> <span class=\"traffic-light green\"></span> </div>"];
	n.lang && r.push(`<div class="code-lang">${e.toUpperCase()}</div>`), r.push("</div>");
	let i = r.join(""), a = ["<div class=\"center\">"];
	n.title && t && a.push(`<div class="code-title">${t}</div>`), a.push("</div>");
	let o = a.join("\n"), s = ["<div class=\"right\">"];
	return n.wrapToggle && s.push("<iconify-icon icon=\"fa6-solid:arrow-down-wide-short\" class=\"toggle-wrap\" title=\"Toggle Wrap\"></iconify-icon>"), n.copyButton && s.push("<div class=\"copy-notice\"></div><iconify-icon icon=\"fa6-solid:paste\" class=\"copy-button\"></iconify-icon>"), s.push("</div>"), `<div class="shiki-tools">${i}${o}${s.join("\n")}</div>`;
}
async function oe(e, t) {
	if (e && t) {
		let a = e();
		await n(i(t), { recursive: !0 }), await r(t, a, "utf8");
	}
}
function se(e, t) {
	let n = (t.match(/<span class="line/g) || []).length, r = e.collapseConfig.enable && n > e.collapseConfig.maxLines;
	return {
		expandButton: r ? "<div class=\"code-expand-btn\"><iconify-icon icon=\"garden:chevron-double-down-fill-16\"></iconify-icon></div>" : "",
		collapseAttrs: r ? ` data-collapsible="true" data-max-lines="${e.collapseConfig.maxLines}" data-total-lines="${n}"` : ""
	};
}
//#endregion
//#region src/index.ts
function ce(e, n) {
	let r = ie(n);
	e.renderer.rules.fence = async (e, n) => {
		let i = e[n];
		if (!i) return "";
		let a = i.content, o = i.info.split(/\s+/)[0] || "", s = i.info.split(/\s+/).slice(1).join(" ");
		if (r.excludes.includes(o)) return `<pre><code class="${o}">${$(a)}</code></pre>`;
		let c = await t(a.replace(/\r?\n$/, ""), {
			lang: r.aliases[o] || o,
			themes: r.themes,
			transformers: r.transformers
		});
		await oe(r.styleToClass.cssGetter, r.styleToClass.css_output_path), c = c.replace(/<pre[^>]*>/, (e) => e.replace(/\s*style\s*=\s*"[^"]*"\s*tabindex="0"/, ""));
		let l = ae(o || "", s || "", r.toolbarItems), { expandButton: u, collapseAttrs: d } = se(r, c);
		return `<figure class="shiki" ${d}> ${l} ${c}${u} </figure>`;
	}, e.renderer.rules.code_inline = async (e, n, i, a, o) => {
		let s = e[n];
		if (!s) return "";
		let c = s.content.trim(), l = c.match(/^\{(\w+)\}\s+(.+)$/);
		if (l === null) return `<code${o.renderAttrs(s)}>${$(c)}</code>`;
		let [, u, d] = l;
		if (!u || !d) return `<code>${c}</code>`;
		let f = await t(d, {
			lang: u,
			themes: r.themes,
			structure: "inline"
		});
		return `<code${o.renderAttrs(s)}>${f}</code>`;
	};
}
//#endregion
export { ce as default };
