// markdown-it-mathjax3-pro/dist/index.js
"use strict";
const mathjax_js_1 = require("mathjax-full/js/mathjax.js");
const tex_js_1 = require("mathjax-full/js/input/tex.js");
const svg_js_1 = require("mathjax-full/js/output/svg.js");
const chtml_js_1 = require("mathjax-full/js/output/chtml.js");
const liteAdaptor_js_1 = require("mathjax-full/js/adaptors/liteAdaptor.js");
const html_js_1 = require("mathjax-full/js/handlers/html.js");
const AllPackages_js_1 = require("mathjax-full/js/input/tex/AllPackages.js");
const assistive_mml_js_1 = require("mathjax-full/js/a11y/assistive-mml.js");
let inlineMathSeparator = [['$', '$']];
let displayMathSeparator = [['$$', '$$']];
function isValidDelim(state, pos, delimiter = '$') {
    let max = state.posMax, can_open = true, can_close = true;
    const prevChar = pos > 0 ? state.src.charCodeAt(pos - 1) : -1;
    const nextChar = pos + delimiter.length <= max ? state.src.charCodeAt(pos + delimiter.length) : -1;
    if (prevChar === 0x20 ||
        prevChar === 0x09 ||
        (nextChar >= 0x30 && nextChar <= 0x39)) {
        can_close = false;
    }
    if (nextChar === 0x20 || nextChar === 0x09) {
        can_open = false;
    }
    return {
        can_open: can_open,
        can_close: can_close,
    };
}
function math_inline(state, silent) {
    let matchedDelimiter = null;
    for (const [openDelim, closeDelim] of inlineMathSeparator) {
        if (state.src.slice(state.pos, state.pos + openDelim.length) === openDelim) {
            matchedDelimiter = [openDelim, closeDelim];
            break;
        }
    }
    if (!matchedDelimiter) {
        return false;
    }
    const [openDelim, closeDelim] = matchedDelimiter;
    let res = isValidDelim(state, state.pos, openDelim);
    if (!res.can_open) {
        if (!silent) {
            state.pending += openDelim;
        }
        state.pos += openDelim.length;
        return true;
    }
    const start = state.pos + openDelim.length;
    let match = start;
    while ((match = state.src.indexOf(closeDelim, match)) !== -1) {
        let pos = match - 1;
        while (state.src[pos] === "\\") {
            pos -= 1;
        }
        if ((match - pos) % 2 == 1) {
            break;
        }
        match += 1;
    }
    if (match === -1) {
        if (!silent) {
            state.pending += openDelim;
        }
        state.pos = start;
        return true;
    }
    if (match - start === 0) {
        if (!silent) {
            state.pending += openDelim + closeDelim;
        }
        state.pos = start + closeDelim.length;
        return true;
    }
    res = isValidDelim(state, match, closeDelim);
    if (!res.can_close) {
        if (!silent) {
            state.pending += openDelim;
        }
        state.pos = start;
        return true;
    }
    if (!silent) {
        const token = state.push("math_inline", "math", 0);
        token.markup = openDelim;
        token.content = state.src.slice(start, match);
    }
    state.pos = match + closeDelim.length;
    return true;
}
function math_block(state, start, end, silent) {
    let next, lastPos;
    let found = false, pos = state.bMarks[start] + state.tShift[start], max = state.eMarks[start], lastLine = "";
    let matchedDelimiter = null;
    for (const [openDelim, closeDelim] of displayMathSeparator) {
        if (state.src.slice(pos, pos + openDelim.length) === openDelim) {
            matchedDelimiter = [openDelim, closeDelim];
            break;
        }
    }
    if (!matchedDelimiter) {
        return false;
    }
    const [openDelim, closeDelim] = matchedDelimiter;
    if (pos + openDelim.length > max) {
        return false;
    }
    pos += openDelim.length;
    let firstLine = state.src.slice(pos, max);
    if (silent) {
        return true;
    }
    if (firstLine.trim().slice(-closeDelim.length) === closeDelim) {
        firstLine = firstLine.trim().slice(0, -closeDelim.length);
        found = true;
    }
    for (next = start; !found;) {
        next++;
        if (next >= end) {
            break;
        }
        pos = state.bMarks[next] + state.tShift[next];
        max = state.eMarks[next];
        if (pos < max && state.tShift[next] < state.blkIndent) {
            break;
        }
        if (state.src.slice(pos, max).trim().slice(-closeDelim.length) === closeDelim) {
            lastPos = state.src.slice(0, max).lastIndexOf(closeDelim);
            lastLine = state.src.slice(pos, lastPos);
            found = true;
        }
    }
    state.line = next + 1;
    const token = state.push("math_block", "math", 0);
    token.block = true;
    token.content =
        (firstLine && firstLine.trim() ? firstLine + "\n" : "") +
            state.getLines(start + 1, next, state.tShift[start], true) +
            (lastLine && lastLine.trim() ? lastLine : "");
    token.map = [start, state.line];
    token.markup = openDelim;
    return true;
}
function processMath(state, documentOptions) {
    const mathTokens = [];
    function walk(tokens) {
        if (!tokens)
            return;
        for (const token of tokens) {
            if (token.type === 'math_inline' || token.type === 'math_block') {
                mathTokens.push(token);
            }
            if (token.children) {
                walk(token.children);
            }
        }
    }
    walk(state.tokens);
    if (mathTokens.length === 0) {
        return;
    }
    const formulas = mathTokens.map(token => token.content);
    const displayModes = mathTokens.map(token => token.type === 'math_block');
    const adaptor = (0, liteAdaptor_js_1.liteAdaptor)();
    const handler = (0, html_js_1.RegisterHTMLHandler)(adaptor);
    (0, assistive_mml_js_1.AssistiveMmlHandler)(handler);
    const combinedContent = formulas.map((content, i) => {
        if (displayModes[i]) {
            const [openDelim, closeDelim] = displayMathSeparator[0];
            return `${openDelim}${content}${closeDelim}`;
        }
        else {
            const [openDelim, closeDelim] = inlineMathSeparator[0];
            return `${openDelim}${content}${closeDelim}`;
        }
    }).join('');
    const mathDocument = mathjax_js_1.mathjax.document(combinedContent, documentOptions);
    mathDocument.render();
    const stylesheet = adaptor.outerHTML(documentOptions.OutputJax.styleSheet(mathDocument));
    state.env.mathjax_stylesheet = extractCssFromStyleTag(stylesheet);
    const body = adaptor.body(mathDocument.document);
    const renderedHtmls = adaptor.childNodes(body).map(node => adaptor.outerHTML(node));
    mathTokens.forEach((token, i) => {
        token.type = displayModes[i] ? 'html_block' : 'html_inline';
        token.content = renderedHtmls[i];
    });
}
function extractCssFromStyleTag(str) {
    const match = str.match(/<style[^>]*>([\s\S]*?)<\/style>/i);
    return match ? match[1] : str;
}
function plugin(md, options) {
    if (options === null || options === void 0 ? void 0 : options.user_side) {
        let tex = options === null || options === void 0 ? void 0 : options.tex;
        md.core.ruler.push('inject_content', (state) => {
            if (!state.env.frontmatter)
                state.env.frontmatter = {};
            let mathjax_options = (options === null || options === void 0 ? void 0 : options.mathjax_options) || {};
            state.env.frontmatter.inject_content = [
                {
                    type: 'script',
                    contribution: { type: 'text/javascript' },
                    content: `window.MathJax = {
                                    startup: { 
                                        ready: () => {
                                            MathJax.startup.defaultReady();
                                            console.log('MathJax loaded');
                                        }
                                    },
                                    tex: ${JSON.stringify({
                        inlineMath: [['$', '$']],
                        displayMath: [['$$', '$$']]
                    }, tex)},
                                    options: ${JSON.stringify(Object.assign({
                        enableMenu: true,
                    }, mathjax_options))}
                            };`
                },
                {
                    type: 'script',
                    contribution: {
                        async: true,
                        src: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
                        id: 'MathJax-script'
                    },
                }
            ];
            return true;
        });
    }
    else {
        inlineMathSeparator = (options === null || options === void 0 ? void 0 : options.tex.inlineMath) || inlineMathSeparator;
        displayMathSeparator = (options === null || options === void 0 ? void 0 : options.tex.displayMath) || displayMathSeparator;
        const documentOptions = {
            InputJax: new tex_js_1.TeX(Object.assign({
                packages: AllPackages_js_1.AllPackages,
                inlineMath: inlineMathSeparator,
                displayMath: displayMathSeparator
            }, options === null || options === void 0 ? void 0 : options.tex)),
            OutputJax: (options === null || options === void 0 ? void 0 : options.svg)
                ? new svg_js_1.SVG(Object.assign({
                    fontCache: 'local',
                    displayAlign: 'center'
                }, options.svg))
                : new chtml_js_1.CHTML(Object.assign({
                    fontURL: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/output/chtml/fonts/woff-v2'
                }, options === null || options === void 0 ? void 0 : options.chtml))
        };
        md.inline.ruler.before("text", "math_inline", math_inline);
        md.block.ruler.after("blockquote", "math_block", math_block, {
            alt: ["paragraph", "reference", "blockquote", "list"],
        });
        md.core.ruler.push('mathjax_render', (state) => {
            state.env.mathjax_stylesheet = null;
            processMath(state, documentOptions);
            return true;
        });
        md.core.ruler.push('inject_content', (state) => {
            if (state.env && state.env.mathjax_stylesheet) {
                if (!state.env.frontmatter)
                    state.env.frontmatter = {};
                state.env.frontmatter.inject_content = [
                    {
                        type: 'style',
                        contribution: { type: 'text/css' },
                        content: state.env.mathjax_stylesheet
                    }
                ];
            }
            return true;
        });
    }
}
plugin.default = plugin;
plugin.displayName = 'MarkdownItMathJaX3PRO';
module.exports = plugin;
